crates/bevy_ecs/macros/src/fetch.rs

```diff
@@ -0,0 +1,577 @@
+
+                // We generate a readonly assertion for every struct member.
+                fn assert_all #impl_generics () #where_clause {
+                    #(assert_readonly::<<#field_types as #path::query::WorldQuery>::Fetch>();)*
+                }
+            };
+        }
+    };
+
+    let tokens = TokenStream::from(quote! {
+        #fetch_impl
+
+        #state_impl
+
+        #read_only_impl
+
+        impl #impl_generics #path::query::WorldQuery for #struct_name #ty_generics #where_clause {
+            type Fetch = #fetch_struct_name #ty_generics;
+            type State = #state_struct_name #ty_generics;
+            type ReadOnlyFetch = #read_only_fetch_struct_name #ty_generics;
+        }
+
+        /// SAFETY: each item in the struct is read only
+        unsafe impl #impl_generics #path::query::ReadOnlyFetch for #read_only_fetch_struct_name #ty_generics #where_clause {}
+
+        // The original struct will most likely be left unused. As we don't want our users having
+        // to specify `#[allow(dead_code)]` for their custom queries, we are using this cursed
+        // workaround.
+        #[allow(dead_code)]
+        const _: () = {
+            fn dead_code_workaround #impl_generics (q: #struct_name #ty_generics) #where_clause {
+                #(q.#field_idents;)*
+                #(q.#ignored_field_idents;)*
+            }
+        };
+    });
+    tokens
+}
+
+struct WorldQueryFieldInfo {
+    /// The original field type.
+    field_type: Type,
+    /// The same as `query_type` but with `'fetch` lifetime.
+    fetch_init_type: Type,
+    /// Has `#[fetch(ignore)]` or `#[filter_fetch(ignore)]` attribute.
+    is_ignored: bool,
+    /// All field attributes except for `world_query` ones.
+    attrs: Vec<Attribute>,
+}
+
+fn read_world_query_field_info(
+    field: &Field,
+    world_lifetime: &Lifetime,
+    fetch_lifetime: &Lifetime,
+) -> WorldQueryFieldInfo {
+    let is_ignored = field
+        .attrs
+        .iter()
+        .find(|attr| {
+            attr.path
+                .get_ident()
+                .map_or(false, |ident| ident == WORLD_QUERY_ATTRIBUTE_NAME)
+        })
+        .map_or(false, |attr| {
+            let mut is_ignored = false;
+            attr.parse_args_with(|input: ParseStream| {
+                if input
+                    .parse::<Option<field_attr_keywords::ignore>>()?
+                    .is_some()
+                {
+                    is_ignored = true;
+                }
+                Ok(())
+            })
+            .unwrap_or_else(|_| {
+                panic!("Invalid `{}` attribute format", WORLD_QUERY_ATTRIBUTE_NAME)
+            });
+
+            is_ignored
+        });
+
+    let attrs = field
+        .attrs
+        .iter()
+        .cloned()
+        .filter(|attr| {
+            attr.path
+                .get_ident()
+                .map_or(true, |ident| ident != WORLD_QUERY_ATTRIBUTE_NAME)
+        })
```

Comment:
This needlesly clones a attribute, thats going to be filtered out anyway.
```suggestion
        .filter(|attr| {
            attr.path
                .get_ident()
                .map_or(true, |ident| ident != WORLD_QUERY_ATTRIBUTE_NAME)
        })
        .cloned()
```